File: mobile/customer-app/app/src/main/java/com/sukashawarma/customer/ui/profile/InformasiAkunViewModel.kt
package com.sukashawarma.customer.ui.profile

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sukashawarma.customer.data.Repository
import com.sukashawarma.customer.data.SessionStore
import com.sukashawarma.customer.data.api.CustomerDto
import com.sukashawarma.customer.data.api.GatewayResult
import com.sukashawarma.customer.ui.components.pesanGalat
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class InformasiAkunState(
    val nama: String = "",
    val email: String? = null,
    val whatsApp: String = "",
    /** Nilai terakhir yang tersimpan di server, untuk tahu ada perubahan atau tidak. */
    val namaTersimpan: String = "",
    val whatsAppTersimpan: String = "",
    val galatNama: String? = null,
    val galatWhatsApp: String? = null,
    val menyimpan: Boolean = false,
    val pesanGalat: String? = null,
    val tersimpan: Boolean = false,
) {
    val adaPerubahan: Boolean
        get() = ProfilForm.rapikanNama(nama) != namaTersimpan ||
            whatsApp.trim() != whatsAppTersimpan
}

class InformasiAkunViewModel(
    private val repository: Repository,
    private val sessionStore: SessionStore,
) : ViewModel() {

    private val _state = MutableStateFlow(dariSesi())
    val state: StateFlow<InformasiAkunState> = _state.asStateFlow()

    init {
        // Sesi lokal bisa basi (nomor diisi saat pesan, nama diubah di HP lain).
        // Segarkan diam-diam; kalau gagal, isian dari sesi tetap dipakai.
        viewModelScope.launch {
            val hasil = repository.profil()
            if (hasil is GatewayResult.Sukses) terapkanDariServer(hasil.data, tandaiTersimpan = false)
        }
    }

    private fun dariSesi(): InformasiAkunState {
        val sesi = sessionStore.baca()
        val nama = sesi?.nama?.trim().orEmpty()
        val wa = ProfilForm.untukIsian(sesi?.telepon)
        return InformasiAkunState(
            nama = nama,
            email = sesi?.email,
            whatsApp = wa,
            namaTersimpan = nama,
            whatsAppTersimpan = wa,
        )
    }

    fun ubahNama(nilai: String) = _state.update {
        it.copy(nama = nilai, galatNama = null, pesanGalat = null, tersimpan = false)
    }

    fun ubahWhatsApp(nilai: String) = _state.update {
        it.copy(whatsApp = nilai, galatWhatsApp = null, pesanGalat = null, tersimpan = false)
    }

    fun simpan() {
        val s = _state.value
        if (s.menyimpan || !s.adaPerubahan) return

        val galatNama = ProfilForm.galatNama(s.nama)
        val galatWa = ProfilForm.galatWhatsApp(s.whatsApp)
        if (galatNama != null || galatWa != null) {
            _state.update { it.copy(galatNama = galatNama, galatWhatsApp = galatWa) }
            return
        }

        val namaBaru = ProfilForm.rapikanNama(s.nama)
        val waBaru = s.whatsApp.trim()
        // Kirim hanya yang berubah: field null tidak disentuh gateway.
        val kirimNama = namaBaru.takeIf { it != s.namaTersimpan }
        val kirimWa = waBaru.takeIf { it != s.whatsAppTersimpan }

        _state.update { it.copy(menyimpan = true, pesanGalat = null, tersimpan = false) }
        viewModelScope.launch {
            when (val hasil = repository.simpanProfil(kirimNama, kirimWa)) {
                is GatewayResult.Sukses -> terapkanDariServer(hasil.data, tandaiTersimpan = true)
                is GatewayResult.Gagal -> _state.update {
                    it.copy(menyimpan = false, pesanGalat = pesanGalat(hasil.error))
                }
            }
        }
    }

    private fun terapkanDariServer(c: CustomerDto, tandaiTersimpan: Boolean) {
        val sesi = sessionStore.baca() ?: return
        sessionStore.simpan(
            token = sesi.token,
            expiresAt = sesi.expiresAt,
            nama = c.name,
            email = c.email ?: sesi.email,
            telepon = c.phone,
        )
        val nama = c.name?.trim().orEmpty()
        val wa = ProfilForm.untukIsian(c.phone)
        _state.update { lama ->
            // Penyegaran awal tak boleh menimpa yang sedang diketik pelanggan.
            val sedangMengetik = !tandaiTersimpan && lama.adaPerubahan
            lama.copy(
                nama = if (sedangMengetik) lama.nama else nama,
                whatsApp = if (sedangMengetik) lama.whatsApp else wa,
                email = c.email ?: lama.email,
                namaTersimpan = nama,
                whatsAppTersimpan = wa,
                menyimpan = false,
                tersimpan = tandaiTersimpan,
            )
        }
    }
}

File: mobile/customer-ios-app/SukaShawarma/UI/Profile/InformasiAkunViewModel.swift
import Foundation
import Observation

/// Padanan `InformasiAkunViewModel.kt`: ubah Nama & No WhatsApp. Email terkunci
/// (identitas akun Google untuk masuk).
@MainActor
@Observable
final class InformasiAkunViewModel {
    var nama = "" { didSet { if nama != oldValue { galatNama = nil; bersihkanPesan() } } }
    private(set) var email: String?
    var whatsApp = "" { didSet { if whatsApp != oldValue { galatWhatsApp = nil; bersihkanPesan() } } }
    /// Nilai terakhir yang tersimpan di server, untuk tahu ada perubahan atau tidak.
    private(set) var namaTersimpan = ""
    private(set) var whatsAppTersimpan = ""
    private(set) var galatNama: String?
    private(set) var galatWhatsApp: String?
    private(set) var menyimpan = false
    private(set) var pesanGalat: String?
    private(set) var tersimpan = false

    var adaPerubahan: Bool {
        ProfilForm.rapikanNama(nama) != namaTersimpan
            || whatsApp.trimmingCharacters(in: .whitespaces) != whatsAppTersimpan
    }

    @ObservationIgnored private let repository: Repository
    @ObservationIgnored private let sessionStore: SessionStore
    /// `didSet` juga terpanggil saat nilai diisi dari server; saat itu pesan
    /// "tersimpan" tak boleh ikut terhapus.
    @ObservationIgnored private var sedangMenerapkan = false

    init(repository: Repository, sessionStore: SessionStore) {
        self.repository = repository
        self.sessionStore = sessionStore
        let sesi = sessionStore.baca()
        let n = sesi?.nama?.trimmingCharacters(in: .whitespaces) ?? ""
        let wa = ProfilForm.untukIsian(sesi?.telepon)
        sedangMenerapkan = true
        nama = n; whatsApp = wa
        sedangMenerapkan = false
        email = sesi?.email
        namaTersimpan = n
        whatsAppTersimpan = wa
    }

    private func bersihkanPesan() {
        guard !sedangMenerapkan else { return }
        pesanGalat = nil
        tersimpan = false
    }

    /// Sesi lokal bisa basi (nomor diisi saat pesan, nama diubah di HP lain).
    /// Segarkan diam-diam; kalau gagal, isian dari sesi tetap dipakai.
    func segarkan() async {
        if case .sukses(let c) = await repository.profil() { terapkanDariServer(c, tandaiTersimpan: false) }
    }

    func simpan() async {
        guard !menyimpan, adaPerubahan else { return }

        let gNama = ProfilForm.galatNama(nama)
        let gWa = ProfilForm.galatWhatsApp(whatsApp)
        if gNama != nil || gWa != nil {
            galatNama = gNama
            galatWhatsApp = gWa
            return
        }

        let namaBaru = ProfilForm.rapikanNama(nama)
        let waBaru = whatsApp.trimmingCharacters(in: .whitespaces)
        // Kirim hanya yang berubah: field nil tidak disentuh gateway. Nomor
        // dikirim apa adanya — gateway yang menormalkan ke 628xxxx.
        let kirimNama = namaBaru != namaTersimpan ? namaBaru : nil
        let kirimWa = waBaru != whatsAppTersimpan ? waBaru : nil

        menyimpan = true
        pesanGalat = nil
        tersimpan = false
        switch await repository.simpanProfil(nama: kirimNama, telepon: kirimWa) {
        case .sukses(let c):
            terapkanDariServer(c, tandaiTersimpan: true)
        case .gagal(let g):
            menyimpan = false
            pesanGalat = SukaShawarma.pesanGalat(g)
        }
    }

    private func terapkanDariServer(_ c: CustomerDto, tandaiTersimpan: Bool) {
        guard let sesi = sessionStore.baca() else { return }
        sessionStore.simpan(token: sesi.token, expiresAt: sesi.expiresAt, nama: c.name,
                            email: c.email ?? sesi.email, telepon: c.phone)
        let n = c.name?.trimmingCharacters(in: .whitespaces) ?? ""
        let wa = ProfilForm.untukIsian(c.phone)
        // Penyegaran awal tak boleh menimpa yang sedang diketik pelanggan.
        let sedangMengetik = !tandaiTersimpan && adaPerubahan
        sedangMenerapkan = true
        if !sedangMengetik { nama = n; whatsApp = wa }
        sedangMenerapkan = false
        email = c.email ?? email
        namaTersimpan = n
        whatsAppTersimpan = wa
        menyimpan = false
        tersimpan = tandaiTersimpan
    }
}
